# Introduction

In [1]:
import numpy as np

In [25]:
a1 = np.arange(1, 11)
print("a1= ", a1)

a1=  [ 1  2  3  4  5  6  7  8  9 10]


In [27]:
print("a1[:-6:-1] = ", a1[:-6:-1])

a1[:-6:-1] =  [10  9  8  7  6]


In this case, the start value is not specified, so it defaults to the end of the list because the step value is negative.

In [28]:
a2 = np.random.randint(10, size=(3, 5))
print("a2 = ", "\n", a2)

a2 =  
 [[9 1 4 6 1]
 [4 7 4 7 0]
 [1 6 1 0 2]]


`np.random.randint(10, ...)` generates random integers from 0 (inclusive) up to 10 (exclusive). So, the possible random numbers are 0, 1, 2, ..., 9.

`size=(3, 5)` is an optional argument that specifies the shape of the output array. Here, it's creating a 2-dimensional array with 3 rows and 5 columns.

In [29]:
print(a2[1:])

[[4 7 4 7 0]
 [1 6 1 0 2]]


In [30]:
print(a2[1: 4])

[[4 7 4 7 0]
 [1 6 1 0 2]]


In Python, when you try to slice a list or an array with a stop index that's beyond the length of the ist/array, Python will return a slice up to the end of the list/array. It doesn't raise an IndexError.

In [31]:
print(a2[1: 4, 1 : 4])

[[7 4 7]
 [6 1 0]]


The first 1:4 refers to the first dimension, which is row
The second 1:4 refers to the second dimension, which is column

In [32]:
print(a2[1: 4: -1, 1: 4: -1])

[]


When the step value is negative, the start index should be larger than the stop index for the slice to include any elements. If the start index is smaller than the stop index, the slice is empty because there's no way to get from the start to the stop by subtracting 1 from the index.

In [33]:
print(a2[:: -1, :: -1])

[[2 0 1 6 1]
 [0 7 4 7 4]
 [1 6 4 1 9]]


The difference between `np.arange(10)` and `np.arange(1, 11)` lies in the starting point and the end point of the created arrays:

`np.arange(10)` generates an array from 0 up to (but not including) 10. So it generates 10 numbers starting from 0. The output will be `[0 1 2 3 4 5 6 7 8 9]`.

`np.arange(1, 11)` generates an array from 1 up to (but not including) 11. So it generates 10 numbers starting from 1. The output will be `[1 2 3 4 5 6 7 8 9 10]`.

In [38]:
a3 = np.arange(1, 10)
print("a3 = ", a3)
print("a3.shape = ", a3.shape)

a3 =  [1 2 3 4 5 6 7 8 9]
a3.shape =  (9,)


`a3.shape = (9,)` means that `a3` is a one-dimensional array containing 9 elements

In [41]:
a4 = a3.reshape((3, 3))
print("a4 = \n", a4)
print("a4.shape = ", a4.shape)

a4 = 
 [[1 2 3]
 [4 5 6]
 [7 8 9]]
a4.shape =  (3, 3)


In [42]:
a5 = a3[np.newaxis, :]
print("a5 = \n", a5)
print("a5.shape = ", a5.shape)

a5 = 
 [[1 2 3 4 5 6 7 8 9]]
a5.shape =  (1, 9)


`a3.shape = (1, 9)` means that `a3` is a two-dimensional array with 1 row and 9 columns.

`a3[np.newaxis, :]` adds a new axis to `a3`, making it a two-dimensional array. The new axis is added as the first dimension.

In [43]:
a6 = a3[: , np.newaxis]
print("a6 = \n", a6)
print("a6.shape = ", a6.shape)

a6 = 
 [[1]
 [2]
 [3]
 [4]
 [5]
 [6]
 [7]
 [8]
 [9]]
a6.shape =  (9, 1)


`a3[:, np.newaxis]` adds a new axis to `a3`, making it a two-dimensional array. The new axis is added as the second dimension.

In [46]:
a7 = np.arange(5)
a8 = np.arange(5, 10)
a9 = np.concatenate([a7, a8])
a10 = np.concatenate([a7, a8, a9])
print("a7 = ", a7)
print("a8 = ", a8)
print("a9 = ", a9)
print("a10 = ", a10)

a7 =  [0 1 2 3 4]
a8 =  [5 6 7 8 9]
a9 =  [0 1 2 3 4 5 6 7 8 9]
a10 =  [0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9]


In [50]:
a11 = np.arange(12).reshape((3, 4))
print("a11 = \n", a11)

a12 = np.arange(100, 112).reshape((3, 4))
print("a12 = \n", a12)

a13 = np.concatenate([a11, a12])
print("a13 = \n", a13)

a14 = np.concatenate([a11, a12], axis=1)
print("a14 = \n", a14)

a11 = 
 [[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]
a12 = 
 [[100 101 102 103]
 [104 105 106 107]
 [108 109 110 111]]
a13 = 
 [[  0   1   2   3]
 [  4   5   6   7]
 [  8   9  10  11]
 [100 101 102 103]
 [104 105 106 107]
 [108 109 110 111]]
a14 = 
 [[  0   1   2   3 100 101 102 103]
 [  4   5   6   7 104 105 106 107]
 [  8   9  10  11 108 109 110 111]]


In [55]:
a15 = np.arange(8).reshape((2, 4))
a16 = np.arange(100, 112).reshape((3, 4))
a17 = np.arange(10, 13).reshape((3, 1))

a18 = np.vstack([a15, a16])
print("a18 = \n", a18)
a19 = np.hstack([a16, a17])
print("a19 = \n", a19)

a18 = 
 [[  0   1   2   3]
 [  4   5   6   7]
 [100 101 102 103]
 [104 105 106 107]
 [108 109 110 111]]
a19 = 
 [[100 101 102 103  10]
 [104 105 106 107  11]
 [108 109 110 111  12]]


In [56]:
a20, a21 = np.split(a18, [3])
print("a20 = \n", a20)
print("a21 = \n", a21)

a20 = 
 [[  0   1   2   3]
 [  4   5   6   7]
 [100 101 102 103]]
a21 = 
 [[104 105 106 107]
 [108 109 110 111]]


In [57]:
a22, a23, a24 = np.split(a18, [2, 3])
print("a22 = \n", a22)
print("a23 = \n", a23)
print("a24 = \n", a24)

a22 = 
 [[0 1 2 3]
 [4 5 6 7]]
a23 = 
 [[100 101 102 103]]
a24 = 
 [[104 105 106 107]
 [108 109 110 111]]


In [59]:
a25, a26, a27 = np.split(a19, [2, 3], axis=1)
print("a25 = \n", a25)
print("a26 = \n", a26)
print("a27 = \n", a27)

a25 = 
 [[100 101]
 [104 105]
 [108 109]]
a26 = 
 [[102]
 [106]
 [110]]
a27 = 
 [[103  10]
 [107  11]
 [111  12]]


In [60]:
a28 = np.arange(20).reshape(4, 5)
print("a28 = \n", a28)

a29, a30 = np.vsplit(a28, [2])
print("a29 = \n", a29)
print("a30 = \n", a30)

a31, a32, a33 = np.hsplit(a28, [2, 4])
print("a31 = \n", a31)
print("a32 = \n", a32)
print("a33 = \n", a33)


a28 = 
 [[ 0  1  2  3  4]
 [ 5  6  7  8  9]
 [10 11 12 13 14]
 [15 16 17 18 19]]
a29 = 
 [[0 1 2 3 4]
 [5 6 7 8 9]]
a30 = 
 [[10 11 12 13 14]
 [15 16 17 18 19]]
a31 = 
 [[ 0  1]
 [ 5  6]
 [10 11]
 [15 16]]
a32 = 
 [[ 2  3]
 [ 7  8]
 [12 13]
 [17 18]]
a33 = 
 [[ 4]
 [ 9]
 [14]
 [19]]


# Mathematical Functions

`np.add`

`np.subtract`

`np.negative`

`np.multiply`

`np.divide`

`np.floor_divide` -- \\\\

`np.power`

`np.mod`

`absolute(abs)`

---
`np.equal`

`np.not_equal`

`np.less`

`np.less_equal`

`np.greater`

`np.greater_equal`

In [3]:
a34 = np.arange(20).reshape([4, 5])
print("a34 = \n", a34)

a34 = 
 [[ 0  1  2  3  4]
 [ 5  6  7  8  9]
 [10 11 12 13 14]
 [15 16 17 18 19]]


In [4]:
print("a34 + 3 = \n", a34 + 3)

a34 + 3 = 
 [[ 3  4  5  6  7]
 [ 8  9 10 11 12]
 [13 14 15 16 17]
 [18 19 20 21 22]]


In [5]:
print("a34 // 5 = \n", a34 // 5)

a34 // 5 = 
 [[0 0 0 0 0]
 [1 1 1 1 1]
 [2 2 2 2 2]
 [3 3 3 3 3]]


In [7]:
print("a34 ** 2 = \n", a34 ** 2)

a34 ** 2 = 
 [[  0   1   4   9  16]
 [ 25  36  49  64  81]
 [100 121 144 169 196]
 [225 256 289 324 361]]


In [8]:
print("-a34 = \n", -a34)

-a34 = 
 [[  0  -1  -2  -3  -4]
 [ -5  -6  -7  -8  -9]
 [-10 -11 -12 -13 -14]
 [-15 -16 -17 -18 -19]]


In [9]:
print("a34 % 3 = \n", a34 % 3)

a34 % 3 = 
 [[0 1 2 0 1]
 [2 0 1 2 0]
 [1 2 0 1 2]
 [0 1 2 0 1]]


### Trigonometric functions

In [10]:
theta = np.linspace(0, np.pi, 5)
print(f"thera = {theta}")

thera = [0.         0.78539816 1.57079633 2.35619449 3.14159265]



The linspace function in NumPy generates a sequence of numbers. It is often used for creating a specific range of values between a start and an end point.

`np.linspace(0, np.pi, 5)` will return an array of 5 evenly spaced numbers over the range from 0 to π.

In [11]:
print(f"sin(theta) = {np.sin(theta)}")

sin(theta) = [0.00000000e+00 7.07106781e-01 1.00000000e+00 7.07106781e-01
 1.22464680e-16]


In [12]:
print(f"cos(theta) = {np.cos(theta)}")

cos(theta) = [ 1.00000000e+00  7.07106781e-01  6.12323400e-17 -7.07106781e-01
 -1.00000000e+00]


In [13]:
print(f"tan(theta) = {np.tan(theta)}")

tan(theta) = [ 0.00000000e+00  1.00000000e+00  1.63312394e+16 -1.00000000e+00
 -1.22464680e-16]


In [14]:
print(f"acrsin(theta) = {np.arcsin(theta)}")

acrsin(theta) = [0.         0.90333911        nan        nan        nan]


C:\Users\GANGZHAO\AppData\Local\Temp\ipykernel_25936\284008034.py:1: RuntimeWarning: invalid value encountered in arcsin
  print(f"acrsin(theta) = {np.arcsin(theta)}")


The numpy function `arcsin` calculates the inverse sine of all values in the input array. The inverse sine is defined only for values in the range from -1 to 1, inclusive.

`np.linspace(0, np.pi, 5)`, it generates an array of 5 numbers between 0 and π (approximately 3.14). However, only the first two numbers are within the range from -1 to 1. The other three numbers are greater than 1, so they are out of the domain of the arcsine function.

That's why we're getting nan (not a number) for those values and a RuntimeWarning about an invalid value encountered in arcsin.

To avoid this, we should ensure that the input to `np.arcsin` is always in the range from -1 to 1. For example, use `np.linspace(-1, 1, 5)` instead of `np.linspace(0, np.pi, 5)`.

In [15]:
print(f"acrcos(theta) = {np.arccos(theta)}")

acrcos(theta) = [1.57079633 0.66745722        nan        nan        nan]


C:\Users\GANGZHAO\AppData\Local\Temp\ipykernel_25936\2674371923.py:1: RuntimeWarning: invalid value encountered in arccos
  print(f"acrcos(theta) = {np.arccos(theta)}")


In [16]:
print(f"acrtan(theta) = {np.arctan(theta)}")

acrtan(theta) = [0.         0.66577375 1.00388482 1.16942282 1.26262726]


### Comparison

In [23]:
a35 = np.random.randint(100, size=(2, 5))
print(f"a35 = \n{a35}")

a35 = 
[[ 1 85 50 32 71]
 [53 36 71 61 11]]


In [25]:
print(f"a35 < 50: \n{a35 < 50}")

a35 < 50: 
[[ True False False  True False]
 [False  True False False  True]]


In [26]:
print(f"a35 == 50: \n{a35 == 50}")

a35 == 50: 
[[False False  True False False]
 [False False False False False]]


### Mask
"Masking" refers to the practice of selecting or modifying parts of an array based on some condition.


This is often done by creating a "mask" array of Boolean values (True/False) that is the same shape as the original array. Each value in the mask corresponds to an element in the original array and indicates whether that element should be selected (True) or not (False).

In [27]:
a36 = np.random.randint(100, size=(3, 5))
print(f"a36 = \n{a36}")

a36 = 
[[12 68 94 52 58]
 [52 74 42 77 52]
 [33 73  5 61 14]]


In [28]:
a37 = a36[a36 < 50]
print(f"a37 = \n{a37}")

a37 = 
[12 42 33  5 14]


In [29]:
print(f"a37.shape = {a37.shape}")

a37.shape = (5,)


### Aggregation

`np.sum`

`np.prod`

`np.mean`

`np.std`

`np.var`

`np.min`

`np.max`

`np.argmin`

`np.argmax`

`np.median`

`np.any`

`np.all`

In [7]:
a38 = np.arange(100)
print(f"a38 = \n{a38}")

print()

a39 = np.add.reduce(a38)
print(f"a39 = {a39}")

a38 = 
[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47
 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71
 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95
 96 97 98 99]

a39 = 4950


`np.add.reduce()` is a NumPy function that performs a specific operation, which is the reduction of elements along a given axis of an array using addition. The `reduce()` function repeatedly applies the addition operation to the elements of the array until a single result is obtained.

In [9]:
# Compare the 3 min values below
a40 = np.random.rand(20)
print(f"a40 = \n{a40}")

m1 = min(a40)
print("Min value: ", m1)

m2 = np.min(a40)
print("Min value: ", m2)

m3 = a40.min()
print("Min value: ", m3)

a40 = 
[0.79115335 0.03218142 0.49296618 0.17871509 0.03340619 0.86638351
 0.98732242 0.16329993 0.00448954 0.28555872 0.62178995 0.44729744
 0.75652866 0.07803022 0.40465178 0.66084044 0.95579867 0.86990226
 0.29354483 0.91733098]
Min value:  0.0044895446826099805
Min value:  0.0044895446826099805
Min value:  0.0044895446826099805


`m1 = min(a40)`:

`min()` is a built-in Python function that returns the minimum value from an iterable, such as a list or an array.

In this case, `min(a40)` finds the minimum value among the elements of the array `a40`.
The result is a single scalar value representing the minimum value.
It is important to note that `min()` is a general Python function and does not specifically belong to NumPy.

`m2 = np.min(a40)`:

`np.min()` is a NumPy function that specifically calculates the minimum value of an array.
In this case, `np.min(a40)` finds the minimum value among the elements of the array `a40`.
The result is a single scalar value representing the minimum value.
The NumPy function `np.min()` is designed to work efficiently with NumPy arrays and has certain additional capabilities compared to the built-in `min()` function.

`m3 = a40.min()`:

`a40.min()` is an instance method of a NumPy array object that performs the same operation as `np.min(a40)`.
It is called directly on the array `a40`.
The result is a single scalar value representing the minimum value.
Using `a40.min()` is a convenient way to directly access the `min()` function from the array object.

### sort vs. argsort

In [13]:
x1 = np.random.randint(20, size=10)
print("Before sorting x1 = ", x1)

# Not modify sorce data
y1 = np.sort(x1)
print("After sorting x1 = ", x1)
print("After sorting y1 = ", y1)

# Modify sorce data
x1.sort()
print("After modifying sorce data and sorting x1 = ", x1)

Before sorting x1 =  [14  4 11  3 15  8 12  2 11  5]
After sorting x1 =  [14  4 11  3 15  8 12  2 11  5]
After sorting y1 =  [ 2  3  4  5  8 11 11 12 14 15]
After modifying sorce data and sorting x1 =  [ 2  3  4  5  8 11 11 12 14 15]


In [14]:
x2 = np.random.randint(20, size=10)
print("Before sorting x2 = ", x2)

# Using argsort
y2 = np.argsort(x2)
print("After argsort y2 = ", y2)

Before sorting x2 =  [10 19  5 18  3 19 15 12  0 10]
After argsort y2 =  [8 4 2 0 9 7 6 3 1 5]


### Array Broadcasting

Broadcasting is a powerful feature in NumPy that allows arrays of different shapes to be used together in arithmetic operations. It provides a convenient way to perform element-wise operations between arrays without the need for explicit looping or expanding the dimensions of the arrays.

In NumPy, for two arrays to be compatible for broadcasting, the following rules must be satisfied:

* Dimensions of the arrays are compatible:  
  The arrays should have the same number of dimensions (or one of them should have fewer dimensions).  
  The size of each dimension in the arrays should either match or one of them should have a size of 1.

  
* Arrays with size 1 in a particular dimension are stretched or "broadcast" to match the size of the other array along that dimension.


When performing operations involving arrays with compatible shapes, broadcasting automatically aligns the arrays and applies the operation element-wise. The smaller array is effectively repeated along the dimensions where broadcasting occurs to match the shape of the larger array.

In [18]:
# a41 is 2-d
a41 = np.arange(15).reshape((3, 5))
print("a41 = \n", a41)
print("a41.shape = ", a41.shape)
# a42 is 1-d
a42 = np.arange(5)
print("a42 = \n", a42)
print("a42.shape = ", a42.shape)
a43 = a41 + a42
print("a43 = \n", a43)
print("a43.shape = ", a43.shape)

a41 = 
 [[ 0  1  2  3  4]
 [ 5  6  7  8  9]
 [10 11 12 13 14]]
a41.shape =  (3, 5)
a42 = 
 [0 1 2 3 4]
a42.shape =  (5,)
a43 = 
 [[ 0  2  4  6  8]
 [ 5  7  9 11 13]
 [10 12 14 16 18]]
a43.shape =  (3, 5)


In [19]:
# a41 is 2-d
a44 = np.arange(5).reshape((5, 1))
print("a44 = \n", a44)
print("a44.shape = ", a44.shape)
# a45 is 1-d
a45 = np.arange(5)
print("a45 = \n", a45)
print("a45.shape = ", a45.shape)
a46 = a44 + a45
print("a46 = \n", a46)
print("a46.shape = ", a46.shape)

a44 = 
 [[0]
 [1]
 [2]
 [3]
 [4]]
a44.shape =  (5, 1)
a45 = 
 [0 1 2 3 4]
a45.shape =  (5,)
a46 = 
 [[0 1 2 3 4]
 [1 2 3 4 5]
 [2 3 4 5 6]
 [3 4 5 6 7]
 [4 5 6 7 8]]
a46.shape =  (5, 5)


So this is how broadcasting work
[0]  
[1]  
[2]  
[3]  
[4]  
copy 5 times along the 1 axis in a44 since it has a size of 1 on 1 axis.

[0 1 2 3 4] copy 5 times along the 0 axis in a45 since it has a size of 1 on 0 axis.

In [21]:
# Unable to broadcast
# a47 and a48 are not compatible even afterr broadcasting
a47 = np.arange(15).reshape((3, 5))
print("a47 = \n", a47)
print("a47.shape = ", a47.shape)
a48 = np.arange(3)
print("a48 = \n", a48)
print("a48.shape = ", a48.shape)
a49 = a47 + a48
print("a49 = \n", a49)
print("a49.shape = ", a49.shape)

a47 = 
 [[ 0  1  2  3  4]
 [ 5  6  7  8  9]
 [10 11 12 13 14]]
a47.shape =  (3, 5)
a48 = 
 [0 1 2]
a48.shape =  (3,)


ValueError: operands could not be broadcast together with shapes (3,5) (3,) 